genomic_data_service/rsid_coordinates_resolver.py

```diff
@@ -11,58 +15,109 @@
+    """
+    This function use catalog api to query SNPs for give region querys.
+    :param region_queries: list of region queries
+    :param source: source of the variants frequency
+    :param maf: minimum allele frequency
+    :return: a list of variants sorted by chrom, start position, ref and alt.
+    there are two APIs to use.
+    If the query is coordiantes, it is more than one base long, we use variantByFrequencySource endpoint.
+    The source and maf have default values, but can be changed by user input.
+    Otherwise, we use variants endpoint.
+    Those two endpoints return all types of variants, so we need to filter for only SNPs.
+    The max limit for the two endpoints is 500.
+    Notification need to be added when:
+    1. the region query is not in the valid format(only coordinates, rsid, spdi and hgvs is allowed).
+    2. the start and end are the same.
+    3. no known variants matching query coordinates found.
+    If the coordinates is one base long, even though no viariants are found, it will not generate notification.
+    Instead, we will still add this coordinates to variants list.
+    """
+    region_queries = list(set(region_queries))
+    notifications = {}
+    query_coordinates = []
+    variants = []
+    api_base = CATALOG_API_VARIANTS
+    api = ''
+    for region_query in region_queries:
+        is_single_base = False
+        # example of region_query: chr1:10000-10001
+        if re.match(r'^(chr[1-9]|chr1[0-9]|chr2[0-2]|chrx|chry)(?:\s+|:)(\d+)(?:\s+|-)(\d+)$', region_query):
+            chrom = region_query.split(':')[0]
+            start_end = region_query.split(':')[-1].split('-')
+            start = int(start_end[0])
+            end = int(start_end[1])
+            if end - start <= 0:
+                notifications[region_query] = (
+                    'Failed: coordinates start should be smaller than coordinates end.'
+                )
+                continue
+            if end - start > 1:
+                api_base = CATALOG_API_FREQ
+                api = f'{api_base}&region={region_query}&source={source}&minimum_af={maf}'
+            else:
+                is_single_base = True
+                api = api_base + '&region={}'.format(region_query)
+
+        # example of region_query: rs4970774
+        elif re.match(r'^rs\d+$', region_query):
+            api = api_base + '&rsid={}'.format(region_query)
+        # example of region_query: NC_000001.11:109726205:A:T
+        elif re.match(r'^NC_\d{6}\.\d{1,2}:\d+:\w:\w$', region_query):
+            api = api_base + '&spdi={}'.format(region_query)
+        # example of region_query: NC_000001.11:g.109726206A>T
+        elif re.match(r'^NC_\d{6}\.\d{1,2}:g\.\d+\w>\w$', region_query):
+            api = api_base + '&hgvs={}'.format(region_query)
+        else:
+            notifications[region_query] = 'Failed: invalid region input'
+            continue
+        logging.info(f'api: {api}')
+        res = requests.get(api).json()
```

Comment:
this line may raise many exceptions... should we include an error message for backend failure such as "try again soon" for example?

Reply:
sounds good. Will add it!